classes/persistent/period.php

```diff
@@ -82,4 +84,40 @@ protected function validate_enddate($value) {
 
         return $startdate < $value;
     }
+
+    /**
+     * Check that the start and end dates don't overlap.
+     *
+     * @throws coding_exception
+     */
+    protected function before_validate() {
+        if (self::overlaps($this)) {
+            throw new coding_exception('Overlap');
```

Comment:
This is a pretty brutal error, I'm not 100% sure how the validation should work between the form object and the persistent object, but I think the validation should be done at the form level so you get a nice error message in context linked to the right field. Possibly there is a way to leverage the persistent validation from inside the form validation but I'm not sure.

If this gets non trivial then I'm happy to just move all the validation to the form instead

Reply:
Some docs here that might be relevant

https://docs.moodle.org/dev/Persistent_form#Using_the_form